src/audit/routes/maintain.py

```diff
@@ -0,0 +1,135 @@
+
+    According to this doc
+    https://wiki.postgresql.org/wiki/INSERT_RETURNING_vs_Partitioning, we could
+    implement a workaround. But since we do not currently need GINO to return
+    the inserted statement, and the row does get inserted before the exception
+    is raised, we just catch and ignore it.
+
+    TODO Once a newer version of GINO is released, upgrade and catch
+    `NoSuchRowError` instead of `AttributeError`.
+    """
+    try:
+        await CATEGORY_TO_MODEL_CLASS[category].create(**data)
+    except AttributeError:
+        pass
+
+
+@router.post("/log/presigned_url", status_code=HTTP_201_CREATED)
+async def create_presigned_url_log(
+    body: CreatePresignedUrlLogInput,
+    background_tasks: BackgroundTasks,
+    auth=Depends(Auth),
+) -> dict:
+    """
+    Create a new `presigned_url` audit log.
+
+    This endpoint does not include any authorization checks, but it is not
+    exposed and is only meant for internal use.
+
+    If the timestamp is omitted from the request body, the current date and
+    time will be used.
+
+    The response is returned _before_ inserting the new audit log in the
+    database, so that POSTing audit logs does not impact the performance of
+    the caller and audit-service failures are not visible to users.
+    """
+    data = body.dict()
+    logger.debug(f"Creating `presigned_url` audit log. Received body: {data}")
+
+    allowed_actions = ["download", "upload"]
+    if data["action"] not in allowed_actions:
+        raise HTTPException(
+            HTTP_400_BAD_REQUEST,
+            f"Action '{data['action']}' is not allowed ({allowed_actions})",
+        )
+
+    # take a timestamp as input, store a datetime in the database
+    if data["timestamp"]:
+        data["timestamp"] = datetime.fromtimestamp(data["timestamp"])
+    else:
+        # timestamp=now is automatically added to rows without timestamp,
+        # but we have to remove the key from the data dict
+        del data["timestamp"]
+
+    background_tasks.add_task(insert_row, "presigned_url", data)
+
+
+@router.post("/log/login", status_code=HTTP_201_CREATED)
+async def create_login_log(
+    body: CreateLoginLogInput,
+    background_tasks: BackgroundTasks,
+    auth=Depends(Auth),
+) -> dict:
+    """
+    Create a new `login` audit log.
+
+    This endpoint does not include any authorization checks, but it is not
+    exposed and is only meant for internal use.
+
+    If the timestamp is omitted from the request body, the current date and
+    time will be used.
+
+    The response is returned _before_ inserting the new audit log in the
+    database, so that POSTing audit logs does not impact the performance of
+    the caller and audit-service failures are not visible to users.
+    """
+    data = body.dict()
+    logger.debug(f"Creating `login` audit log. Received body: {data}")
+
+    # take a timestamp as input, store a datetime in the database
+    if data["timestamp"]:
```

Comment:
Would `has_key()` be more explicit?

Reply:
the key is always present because of https://github.com/uc-cdis/audit-service/blob/10d16f7822103fb9af195bcf049a23117b6191fe/src/audit/models.py#L39

Reply:
✔️ Thanks!